src/handler.ts

```diff
@@ -42,27 +45,43 @@ interface Matched {
  * @param projects Sentry extension projects configurations.
  * @return Sentry projectID this document reports to.
  */
-export function matchSentryProject(params: Params, projects: SentryProject[]): Matched | null {
-    if (!projects || !params.repo || !params.file) {
+export function matchSentryProject(params: Params | null, projects: SentryProject[]): Matched | null {
+    if (!projects || !params || !params.repo || !params.file) {
         return null
     }
+
+    let missingConfigs
+
     // Check if a Sentry project is associated with this document's repository and/or file and retrieve the project.
     // TODO: Handle the null case instead of using a non-null assertion !
     // TODO: Handle cases where the wrong project is matched due to similar repo name,
     // e.g. `sourcegraph-jetbrains` repo will match the `sourcegraph` project
-    for (const project of projects) {
-        const missingConfigs = findEmptyConfigs(project)
+    for (const [index, project] of projects.entries()) {
+        missingConfigs = missingConfigs
+            ? missingConfigs.concat(findEmptyConfigs(project, 'project', index))
+            : findEmptyConfigs(project, 'project', index)
 
-        for (const filter of project.filters) {
-            // both repository and file match
-            if (filter.files && !matchesFile(filter.files, params.file)) {
-                break
-            }
+        if (project.filters) {
+            for (const filter of project.filters) {
+                if (filter.files && !matchesFile(filter.files, params.file)) {
+                    break
+                }
 
-            if (filter.repositories && !matchesRepository(filter.repositories, params.repo)) {
-                break
+                if (
+                    filter.repositories &&
+                    filter.repositories.length > 0 &&
+                    !matchesRepository(filter.repositories, params.repo)
+                ) {
+                    break
+                }
+
+                // both repository and file match
+                return { project, missingConfigs }
             }
-            return { project, missingConfigs }
+            // if there is only one project in the settings and no filters,
+            // always match with that Sentry project
+        } else if (projects.length === 1) {
+            return { project, missingConfigs: [] }
         }
```

Comment:
The use of looping, conditionals, special cases, and guards make this difficult for me to follow.

Some ideas that come to mind:

- Try replacing the loops + `if`s with `filter()`s
- Try pushing the `filter.repositories.length > 0` check down into `matchesRepository()`
- Try moving all special cases to the top of the function (not inside loop bodies)
- See if you can put it all into one expression: `return { project: projects.find(...), missingConfigs: projects.map(...) }`. I see that `project` and `missingConfigs` are independent of each other, but they are referred to in the same loop.

Reply:
Will be addressing those suggestions in a new PR and will use this issue tracker: https://github.com/sourcegraph/sourcegraph-sentry/issues/56